boards/shared/Src/Io/Io_SharedFrequencyOnlyPwmInput.c

```diff
@@ -0,0 +1,127 @@
+#include <main.h>
+#include <math.h>
+#include <assert.h>
+#include "Io_SharedFrequencyOnlyPwmInput.h"
+
+#define MAX_TIMER_PERIOD_ELAPSED_COUNT 2
+#define AUTO_RELOAD_REG 65535
+
+struct FreqOnlyPwmInput
+{
+    float              frequency_hz;
+    TIM_HandleTypeDef *htim;
+    float              timer_frequency_hz;
+    uint32_t           rising_edge_tim_channel;
+    bool               first_tick;
+    uint32_t           curr_rising_edge;
+    uint32_t           prev_rising_edge;
+};
+
+/**
+ * Set the frequency for the given PWM input
+ * @param pwm_input: The PWM input to set frequency for
+ * @param frequency_hz: Frequency, in Hz
+ */
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *pwm_input,
+    const float              frequency_hz);
+
+static void Io_SetFrequency(
+    struct FreqOnlyPwmInput *const pwm_input,
+    const float                    frequency_hz)
+{
+    assert(frequency_hz >= 0.0f || isnanf(frequency_hz));
+
+    pwm_input->frequency_hz = frequency_hz;
+}
+
+struct FreqOnlyPwmInput *Io_SharedFreqOnlyPwmInput_Create(
+    TIM_HandleTypeDef *htim,
+    float              timer_frequency_hz,
+    uint32_t           rising_edge_timer_channel)
+{
+    assert(htim != NULL);
+    static struct FreqOnlyPwmInput pwm_inputs[MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS];
+    static size_t                  alloc_index = 0;
+
+    assert(alloc_index < MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS);
+
+    struct FreqOnlyPwmInput *const pwm_input = &pwm_inputs[alloc_index++];
+
+    pwm_input->htim                    = htim;
+    pwm_input->timer_frequency_hz      = timer_frequency_hz;
+    pwm_input->rising_edge_tim_channel = rising_edge_timer_channel;
+    pwm_input->first_tick              = true;
+
+    HAL_TIM_IC_Start_IT(htim, rising_edge_timer_channel);
+    HAL_TIM_Base_Start_IT(htim);
+
+    return pwm_input;
+}
+
+void Io_SharedFreqOnlyPwmInput_Tick(struct FreqOnlyPwmInput *const pwm_input)
+{
+    if (pwm_input->first_tick)
+    {
+        pwm_input->curr_rising_edge = HAL_TIM_ReadCapturedValue(
+            pwm_input->htim, pwm_input->rising_edge_tim_channel);
+        pwm_input->first_tick = false;
+    }
+    else
+    {
+        pwm_input->first_tick       = true;
```

Comment:
Why would we set first_tick back to true?

Reply:
fixed